### PR TITLE
Show account on discord rich presence

### PR DESCRIPTION
**Pull request by axellse**
Implements #2044

Shows your roblox avatar and username as the small image on your discord profile. Only shows if you have enabled "Allow activity joining" in the Bloxstrap menu.

Heres a screenshot of what it looks like on your Discord profile: 
image

### CODE REVIEW

**Comment by axellse** [comment]
Also sorry for the commit history, im on a chromebook rn and im using the github online editor thing

File: Bloxstrap/Resources/Strings.resx [hunk 1]
```diff
@@ -761,6 +761,12 @@ Selecting 'No' will ignore this warning and continue installation.</value>
   <data name="Menu.Integrations.AllowActivityJoining.Title" xml:space="preserve">
     <value>Allow activity joining</value>
   </data>
+  <data name="Menu.Integrations.ShowAccountOnProfile.Description" xml:space="preserve">
+    <value>Shows what Roblox account your using on your Discord profile.</value>
```
Comment (Flikter) [inline Bloxstrap/Resources/Strings.resx:765]: I suggest changing this to: "Show which Roblox account you are using on your Discord profile."

Comment (axellse) [reply Bloxstrap/Resources/Strings.resx:765]: do you think the original sounds weird or why do you think i should change it? 👍

Comment (Flikter) [reply Bloxstrap/Resources/Strings.resx:765]: I believe replacing "what" with "which" sounds better. The part "Roblox account **your** using" is incorrect, it should be "Roblox account **you're** using". I used the uncontracted form of "you're" as I heard it's better for translation accuracy.

Comment (axellse) [reply Bloxstrap/Resources/Strings.resx]: thanks for the feedback, i've changed it. english is not my native language.

**Comment by pizzaboxer** [comment]
Looks good, I can't add this into the next version though since translations are all done already. It'll be merged for the version coming after.

File: Bloxstrap/Resources/Strings.resx (hunk 1 above)
Comment (ShadowCodeX-debug) [reply Bloxstrap/Resources/Strings.resx]: @Flikter You’re is “you are.” Your is a possessive adjective, meaning it shows ownership.

File: Bloxstrap/Integrations/DiscordRichPresence.cs [hunk 2]
```diff
@@ -231,6 +235,31 @@ public async Task<bool> SetCurrentGame()
...
+                {
+                    smallImage = userPfpResponse.Data.ToArray()[0].ImageUrl;
+                    App.Logger.WriteLine(LOG_IDENT, $"Got user thumbnail as {smallImage}");
+                }
+                
+                var userInfoResponse = await Http.GetJson<UserInfoResponse>($"https://users.roblox.com/v1/users/{userId}");
+                if (userInfoResponse is null)
+                {
+                    App.Logger.WriteLine(LOG_IDENT, "Could not get user info!");
+                } else
```
Comment (bluepilledgreat) [inline Bloxstrap/Integrations/DiscordRichPresence.cs:255]: Can you put the bracket and the else statement on separate lines?

**Comment by CubesterYT** [comment]
Oh wow, this is EXACTLY how I imagined it would be! Great work so far! I can't wait to see it in production

**Comment by axellse** [comment]
@bluepilledgreat would you mind submitting approving the changes or submitting another code revirew?

**Comment by Vlad55432** [comment]
Any chance, that this would be merged?

**Review by pizzaboxer — APPROVED** [review]
Looks great, though before I merge it I'll need to implement caching of user info (the same way that universe details are now cached).

**Comment by GalacticInspired** [comment]
noo my secret alt accounts

File: Bloxstrap/Integrations/ActivityWatcher.cs [hunk 3]
```diff
@@ -212,6 +214,21 @@ private void ReadLogEntry(string entry)
...
+
+                    if (match.Groups.Count != 2)
+                    {
+                        App.Logger.WriteLine(LOG_IDENT, $"Failed to assert format for game join load time entry");
+                        App.Logger.WriteLine(LOG_IDENT, entry);
+                        return;                
+                    }
+
+                    Data.UserId = match.Groups[1].Value;
+                    App.Logger.WriteLine(LOG_IDENT, $"Got userid as " + Data.UserId);
```
Comment (bluepilledgreat) [inline Bloxstrap/Integrations/ActivityWatcher.cs:229]: $"Got User ID as {Data.UserId}"

File: Bloxstrap/Integrations/ActivityWatcher.cs [hunk 4]
```diff
@@ -212,6 +214,21 @@ private void ReadLogEntry(string entry)
             {
                 // We are not confirmed to be in a game, but we are in the process of joining one
 
+                if (entry.Contains(GameJoinLoadTimeEntry))
+                {
+                    Match match = Regex.Match(entry, GameJoinLoadTimeEntryPattern);
+
+                    if (match.Groups.Count != 2)
+                    {
+                        App.Logger.WriteLine(LOG_IDENT, $"Failed to assert format for game join load time entry");
```
Comment (bluepilledgreat) [inline Bloxstrap/Integrations/ActivityWatcher.cs:223]: Remove the '$'. $ is the identifier for a formatted string. You are not using formatting here.

File: Bloxstrap/Integrations/ActivityWatcher.cs [hunk 5]
```diff
@@ -19,7 +19,9 @@ public class ActivityWatcher : IDisposable
         private const string GameDisconnectedEntry           = "[FLog::Network] Time to disconnect replication data:";
         private const string GameTeleportingEntry            = "[FLog::SingleSurfaceApp] initiateTeleport";
         private const string GameLeavingEntry                = "[FLog::SingleSurfaceApp] leaveUGCGameInternal";
+        private const string GameJoinLoadTimeEntry = "[FLog::GameJoinLoadTime] Report game_join_loadtime:";
 
+        private const string GameJoinLoadTimeEntryPattern = ", userid:([0-9]+)";
```
Comment (bluepilledgreat) [inline Bloxstrap/Integrations/ActivityWatcher.cs:24]: Add some spaces to match with the other constants.

File: Bloxstrap/Integrations/ActivityWatcher.cs [hunk 6]
```diff
@@ -19,7 +19,9 @@ public class ActivityWatcher : IDisposable
         private const string GameDisconnectedEntry           = "[FLog::Network] Time to disconnect replication data:";
         private const string GameTeleportingEntry            = "[FLog::SingleSurfaceApp] initiateTeleport";
         private const string GameLeavingEntry                = "[FLog::SingleSurfaceApp] leaveUGCGameInternal";
+        private const string GameJoinLoadTimeEntry = "[FLog::GameJoinLoadTime] Report game_join_loadtime:";
```
Comment (bluepilledgreat) [inline Bloxstrap/Integrations/ActivityWatcher.cs:22]: Add some spaces to match with the other constants.

File: Bloxstrap/Integrations/DiscordRichPresence.cs [hunk 7]
```diff
@@ -224,6 +226,31 @@ public async Task<bool> SetCurrentGame()
...
+            if (App.Settings.Prop.AccountShownOnProfile)
+            {
+                var userPfpResponse = await Http.GetJson<ApiArrayResponse<ThumbnailResponse>>($"https://thumbnails.roblox.com/v1/users/avatar-headshot?userIds={activity.UserId}&size=180x180&format=Png&isCircular=false"); //we can remove '-headshot' from the url if we want a full avatar picture
+                if (userPfpResponse is null || !userPfpResponse.Data.Any())
+                {
+                    App.Logger.WriteLine(LOG_IDENT, "Could not get user thumbnail info!");
+                }
+                else
+                {
+                    smallImage = userPfpResponse.Data.ToArray()[0].ImageUrl;
```
Comment (bluepilledgreat) [inline Bloxstrap/Integrations/DiscordRichPresence.cs:238]: Pretty sure you can do `userPfpResponse.Data.First().ImageUrl`

File: Bloxstrap/Integrations/ActivityWatcher.cs [hunk 8]
```diff
@@ -212,6 +214,21 @@ private void ReadLogEntry(string entry)
...
+                    Match match = Regex.Match(entry, GameJoinLoadTimeEntryPattern);
+
+                    if (match.Groups.Count != 2)
+                    {
+                        App.Logger.WriteLine(LOG_IDENT, $"Failed to assert format for game join load time entry");
+                        App.Logger.WriteLine(LOG_IDENT, entry);
+                        return;                
+                    }
+
+                    Data.UserId = match.Groups[1].Value;
```
Comment (bluepilledgreat) [inline Bloxstrap/Integrations/ActivityWatcher.cs:228]: Convert the user ID to a ulong instead of storing it as a string.

Comment (bluepilledgreat) [reply Bloxstrap/Integrations/ActivityWatcher.cs:228]: Use ulong.TryParse.

**Review by bluepilledgreat — CHANGES_REQUESTED** [review]

File: Bloxstrap/Integrations/ActivityWatcher.cs [hunk 9]
```diff
@@ -217,13 +217,15 @@ private void ReadLogEntry(string entry)
...
                     {
-                        App.Logger.WriteLine(LOG_IDENT, $"Failed to assert format for game join load time entry");
+                        App.Logger.WriteLine(LOG_IDENT, "Failed to assert format for game join load time entry");
                         App.Logger.WriteLine(LOG_IDENT, entry);
                         return;                
                     }
 
-                    Data.UserId = match.Groups[1].Value;
-                    App.Logger.WriteLine(LOG_IDENT, $"Got userid as " + Data.UserId);
+                    UInt64.TryParse(match.Groups[1].Value, out ulong result);
```
Comment (bluepilledgreat) [inline Bloxstrap/Integrations/ActivityWatcher.cs:225]: Handle failure. TryParse returns a boolean. False = failed to parse.

**Review by bluepilledgreat — CHANGES_REQUESTED** [review]

**Review by bluepilledgreat — APPROVED** [review]

**Comment by pizzaboxer** [comment]
I'll take it from here, thank you Axell.

**Comment by CubesterYT** [comment]
LETS GO!

**Comment by Vlad55432** [comment]
YAYYYy!

**Pull request merged**